// prettier-plugin-solidity/dist/slang-nodes/ExponentiationExpression.js
import { doc } from 'prettier';
import { NonterminalKind } from '@nomicfoundation/slang/cst';
import { createBinaryOperationPrinter } from '../slang-printers/create-binary-operation-printer.js';
import { binaryIndentRulesBuilder } from '../slang-printers/print-binary-operation.js';
import { createHugFunction } from '../slang-utils/create-hug-function.js';
import { createKindCheckFunction } from '../slang-utils/create-kind-check-function.js';
import { getNodeMetadata, updateMetadata } from '../slang-utils/metadata.js';
import { Expression } from './Expression.js';
const { group } = doc.builders;
const tryToHug = createHugFunction(['**']);
const shouldIndent = createKindCheckFunction([
    NonterminalKind.MultiplicativeExpression,
    NonterminalKind.AdditiveExpression,
    NonterminalKind.ShiftExpression,
    NonterminalKind.BitwiseAndExpression,
    NonterminalKind.BitwiseOrExpression,
    NonterminalKind.BitwiseXorExpression,
    NonterminalKind.InequalityExpression,
    NonterminalKind.EqualityExpression,
    NonterminalKind.AndExpression,
    NonterminalKind.OrExpression
]);
const printExponentiationExpression = createBinaryOperationPrinter(() => (document) => group(document), // always group
binaryIndentRulesBuilder(shouldIndent) // indent as a binary operation with some exceptions
);
export class ExponentiationExpression {
    constructor(ast, options) {
        this.kind = NonterminalKind.ExponentiationExpression;
        let metadata = getNodeMetadata(ast);
        this.leftOperand = new Expression(ast.leftOperand, options);
        this.operator = ast.operator.unparse();
        this.rightOperand = new Expression(ast.rightOperand, options);
        metadata = updateMetadata(metadata, [this.leftOperand, this.rightOperand]);
        this.comments = metadata.comments;
        this.loc = metadata.loc;
        this.rightOperand = tryToHug(this.rightOperand);
        this.leftOperand = tryToHug(this.leftOperand);
    }
    print(path, print, options) {
        return printExponentiationExpression(this, path, print, options);
    }
}
